src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc

```diff
@@ -0,0 +1,500 @@
+void LogicalAudioSensorPlugin::PreUpdate(const UpdateInfo &_info,
+                EntityComponentManager &_ecm)
+{
+  if (this->dataPtr->sourceEntity != kNullEntity)
+  {
+    auto startTime = _info.simTime - _info.dt;
+
+    auto& playInfo = _ecm.Component<components::LogicalAudioSourcePlayInfo>(
+        this->dataPtr->sourceEntity)->Data();
+
+    // configure the source's play information before starting the simulation
+    if (this->dataPtr->firstTime)
+    {
+      playInfo.startTime = startTime;
+      this->dataPtr->firstTime = false;
+    }
+
+    // start playing a source if the play source service was called
+    std::unique_lock<std::mutex> play_lock(this->dataPtr->playSourceMutex);
+    if (this->dataPtr->playSource)
+    {
+      playInfo.playing = true;
+      playInfo.startTime = startTime;
+      this->dataPtr->playSource = false;
+    }
+    play_lock.unlock();
+
+    // stop playing a source if the stop source service was called
+    std::unique_lock<std::mutex> stop_lock(this->dataPtr->stopSourceMutex);
+    if (this->dataPtr->stopSource)
+    {
+      playInfo.playing = false;
+      this->dataPtr->stopSource = false;
+    }
+    stop_lock.unlock();
+
+    // stop playing a source if the play duration has been exceeded
+    if (this->dataPtr->DurationExceeded(_info, playInfo))
+      playInfo.playing = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void LogicalAudioSensorPlugin::PostUpdate(const UpdateInfo &_info,
+                const EntityComponentManager &_ecm)
+{
+  // check to see which sources a microphone can hear
+  if (this->dataPtr->micEntity != kNullEntity)
+  {
+    const auto micPose = worldPose(this->dataPtr->micEntity, _ecm);
+    const auto micInfo = _ecm.Component<components::LogicalMicrophone>(
+        this->dataPtr->micEntity)->Data();
+
+    _ecm.Each<components::LogicalAudioSource,
+              components::LogicalAudioSourcePlayInfo>(
+      [&](const Entity &_entity,
+          const components::LogicalAudioSource *_source,
+          const components::LogicalAudioSourcePlayInfo *_playInfo)
+      {
+        // skip this source if the playing duration has been exceeded
+        // (this source will be stopped in the following PreUpdate call)
+        if (this->dataPtr->DurationExceeded(_info, _playInfo->Data()))
+          return true;
+
+        const auto sourcePose = worldPose(_entity, _ecm);
+        const auto vol = logical_audio::ComputeVolume(
+            _playInfo->Data().playing,
+            _source->Data().attFunc,
+            _source->Data().attShape,
+            _source->Data().emissionVolume,
+            _source->Data().innerRadius,
+            _source->Data().falloffDistance,
+            sourcePose,
+            micPose);
+
+        if (logical_audio::Detect(vol, micInfo.volumeDetectionThreshold))
+        {
+          // TODO(adlarkin) let others know that a detection occurred,
+          // perhaps by publishing a message to a topic?
```

Comment:
Does everyone think that publishing a message is a good way to notify others of a microphone detection, or is there a better way of doing this?

Reply:
yeah I think we should publish the result over a topic. Once that's added, I recommend adding a simple integration test with a similar setup as the example world to verify the functionality.

Reply:
I can publish a message that contains the microphone ID, the IDs of the sources that the microphone heard, and the volume level at the microphone for each of the detected sources. Would that work?

Reply:
looking at the [GasEmitterDetectorPlugin](https://github.com/osrf/subt/blob/master/subt_ign/src/GasEmitterDetectorPlugin.cc#L151) as example, it creates a a topic per detector and publishes msg to that topic. So you could do something similar, i.e. a topic per microphone, and publish a floating point number (volume) to that topic.

Reply:
I'll base it on that :+1: I think it would also be worth adding the IDs of the source(s) a microphone detects in case different different sources trigger different actions by the agent perceiving the sources.

Reply:
I went ahead and added a microphone detection publisher in https://github.com/ignitionrobotics/ign-gazebo/pull/401/commits/f35c133ef7e22c432ee2db786f74c9a814a03bab. As you can see, it publishes an `igniton::msgs::Double` message that contains the volume level of the source that was detected, along with the detected source's ID in the message header:

https://github.com/ignitionrobotics/ign-gazebo/blob/f35c133ef7e22c432ee2db786f74c9a814a03bab/src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc#L241-L256